### ISSUE TITLE
[SSR] Seroval can't serialize axios error in createResource and breaks the app

### ISSUE THREAD

**Issue by mattersj** [solidjs/solid#1919]
### Describe the bug

When `createResource` throws some specific error (in this case it's axios error) that seroval can't handle, the app gets broken with the following error message: "Error: Cannot serialize function without reference ID."

### Your Example Website or App

https://stackblitz.com/edit/solid-ssr-vite-ispxqp?file=src/routes/index.tsx

### Steps to Reproduce the Bug or Issue

0. This bugs can only be reproduced when using SSR.
1. Go to the terminal and observe the described error message.

### Expected behavior

`createResource` should be able to handle any error no matter the type and what place this error was thrown from.
If the error can't serialized for some reason, `createResource` shouldn't break the whole app and probably fallback to something meaningful (default error + message from the original error?).

### Platform

- OS: macOS 12.6.8
- Browser: Chrome
- Version: 118
-

**Comment by lxsmnsyc** [comment]
Seems like `axios.get`'s error has a property that contains a function (moreover, the error object itself is a complex object and has values that seroval cannot serialize, I tried inspecting it myself), this would require the user to intercept the error so that seroval is able to serialize it.
Example:
```js
  const [data] = createResource(async () => {
    try {
      return await axios.get('https://some-random-nonexistent-url.org');
    } catch (error) {
      return new Error(error.message);
    }
  });
```

**Comment by mattersj** [comment]
@lxsmnsyc Yeah that makes sense but I think this inconsistent behavior might not be the best developer experience and sometimes you can't really predict what error would be thrown from a library and whether seroval could handle that error or just fail.
Just imagine a case where some third party library throws a simple error that seroval **can** handle but the other day they decided to add some useful stuff in there that seroval **can't** handle anymore and after update your app is crashed because you didn't predict this beforehand and don't handle your error. Sounds not that great, right?

In additional, `createResource` is supposed to handle errors on its own (that's why we have `error` property) so you can catch them later with `<ErrorBoundary />` somewhere in your code. So it should handle any error in such case, not just a subset of them.

**Comment by lxsmnsyc** [comment]
> supposed to handle errors on its own

It does however allowed errors in createResource has the same restriction as the allowed values it can resolve into. The thing is, SSR serializes the result of createResource, and it cannot serialize any value that it can't handle. It breaks correctness on both server and client.

**Comment by mattersj** [comment]
> allowed errors in createResource has the same restriction as the allowed values it can resolve into.

Yep, I understand that but my point here is that the current developer experience can definitely be improved, at the very least we can provide more useful error with some suggestions if we can't do any better here.
However, I think we still have a few other options to improve the behavior: throwing a custom useful error instead of an original one (and stripping unserializable fields) or maybe forcing a user to handle errors on their own, or maybe something else.
Anyway, we have a work to do here because the current error could be confusing for many people.

**Comment by lxsmnsyc** [comment]
Okay I guess our common ground here is for Solid to intercept Seroval's errors and give a Solid-related warning.

**Comment by mattersj** [comment]
@lxsmnsyc At least yeah, but it would be much more convenient to have something more high-level so you don't have to re-throw a common error because you don't even know beforehand whether an original error could be serialized or not.

**Comment by lxsmnsyc** [comment]
@mattersj whether something is serializable or not is going to be documented in the future for sure, but we can't really change the original intent of the user as it introduces more inconsistencies.

**Comment by ryansolid** [comment]
Ok. Looking at what is a reasonable resolution.

What I'm thinking:

If the value you can't be serialized whether it be promise success or error, the fact it can't serialize has to trump as an error as we can't do anything with the original value anyway. Like it might be an error. It might not be. 

Ideally this error should not crash the server.. if it could be caught in the resource as well then we can just have the normal error boundary behavior.

Other considerations:

More human readable error. If we go that way it probably won't be much clearer. But it is probably doable.

Should non-serializable fields cause a serialization failure or just be skipped. I can tell being skipped is what is desired here, but in the general case that is dangerous because what you put in is not what you get out.


Let me look and see what we can feasibly do here.

**Comment by lxsmnsyc** [comment]
> Should non-serializable fields cause a serialization failure or just be skipped. I can tell being skipped is what is desired here, but in the general case that is dangerous because what you put in is not what you get out.

This is what I'm against, and given Seroval's current structure, this is highly unlikely to happen.

**Comment by ryansolid** [comment]
We added an `onError` to the server render function to collect seroval errors in 1.8.4 but I think it is interesting question on what should happen to the request in these scenarios. We can't really handle this at error boundaries as it is detached, but maybe we could send a 500 response back? I don't know if we should default that or just pass that power completely to the consumer. But at least for now we aren't causing the server to crash to my knowledge.

**Comment by mattersj** [comment]
I haven't had a chance to test 1.8.4 in SolidStart since it still uses Solid 1.8.3 so the server still gets crashed, but I think it'd be great if we could introduce a user-friendly 500 error in these cases explaining why things went wrong and pointing out the exact spot this happened. I'd say Next.js does a good job here:
<img width="522" alt="image" src="https://github.com/solidjs/solid/assets/28510814/08f9f047-9424-459c-9b02-58e00afeeb91">

**Comment by ryansolid** [comment]
Yeah I'm consider `onError` the fix. We can't localize the errorand sort of stuck where we are at I think. If there are new issues that arise around this I'd be interested but for now I'm going to say this issue has run its course.
